Refactor how shared ABI classes are stored

da-woods opened the pull request:
Partial fix for #6613 (I still think there might be at least one case where the USED macros aren't appropriate in coroutine, but I haven't investigated in great detail yet).

Essentially what I'm worried about in this PR is:
* We access the shared ABI types through `CGLOBAL` for whatever Cython module gets imported first.
* The lifetime of those types can be longer than any individual Cython module (because they're shared) but the shared ABI types have no ref-counting linking them back to the module they rely on for their type checks.

Therefore, instead access to the shared types (from within the shared ABI code) goes through the shared module, which has its own module state struct.

Also fixes some data races related to initializing the shared types (if multiple cython modules are being imported from different threads).

da-woods committed:
Creation of shared module

da-woods committed:
Lookup/cleanup implemented

da-woods committed:
Assignment working

da-woods committed:
Ensure that coroutine/generator is consistent

da-woods committed:
Roll back some of the unnecessary changes

da-woods committed:
Bits of small tidy-up

da-woods committed:
Merge remote-tracking branch 'real_origin/master' into shared-abi-refactor-2

da-woods committed:
Cleanup

da-woods committed:
More WIP

da-woods committed:
More WIP

da-woods committed:
Seems to be working

da-woods commented on Cython/Utility/AsyncGen.c:
```diff
@@ -2,26 +2,6 @@
-PyTypeObject *__pyx__PyAsyncGenAThrowType;
-PyTypeObject *__pyx_AsyncGenType;
-
-// Freelists boost performance 6-10%; they also reduce memory
-// fragmentation, as _PyAsyncGenWrappedValue and PyAsyncGenASend
-// are short-living objects that are instantiated for every
-// __anext__ call.
-
-#if CYTHON_USE_FREELISTS
-struct __pyx__PyAsyncGenWrappedValue *__Pyx_ag_value_freelist[_PyAsyncGen_MAXFREELIST];
-int __Pyx_ag_value_freelist_free;
-
-struct __pyx_PyAsyncGenASend *__Pyx_ag_asend_freelist[_PyAsyncGen_MAXFREELIST];
-int __Pyx_ag_asend_freelist_free;
-#endif
```
Moved rather than removed

da-woods commented on Cython/Compiler/ModuleNode.py:
```diff
@@ -2944,11 +2939,8 @@ def generate_module_state_traverse(self, env, code):
         code.putln(f'__Pyx_VISIT_CONST(traverse_module_state->{Naming.empty_tuple});')
         code.putln(f'__Pyx_VISIT_CONST(traverse_module_state->{Naming.empty_bytes});')
         code.putln(f'__Pyx_VISIT_CONST(traverse_module_state->{Naming.empty_unicode});')
-        code.putln('#ifdef __Pyx_CyFunction_USED')
-        code.putln(f'Py_VISIT(traverse_module_state->{Naming.cyfunction_type_cname});')
-        code.putln('#endif')
-        code.putln('#ifdef __Pyx_FusedFunction_USED')
```
Note that only CyFunction and FusedFunction were cleaned up properly - now they all are

da-woods committed:
Fix C compatibility stuff

da-woods committed:
Fix version check

da-woods committed:
Merge remote-tracking branch 'real_origin/master' into shared-abi-refactor-2

da-woods commented:
It looks like anything involving Tempita `py:` is very persistent and is somehow sneaking into other utility code context. That'll need investigating later.

scoder commented:
Wow. I wasn't expecting any further changes in *this* order in the near future any more (certainly not for 3.1), after introducing the shared module.

This seems a huge change with potentially serious performance impact at this point. Type checks should be instantaneous and are often done *because* they are fast and allow quick shortcuts. With your change, they look prohibitively expensive and would probably require a full investigation which of them are still worth it and which already produce a slow-down by just being there.

Honestly, it seems much simpler to ask users to make sure they load at least one persistent module up-front, so that the types don't go away. That usually solves the problem (as I understand it) quite fine.

The problem is also partly reduced by moving shared types to the shared module, although it persists across packages.

And, BTW, as long as users don't loudly request support for unloading Cython modules, why bother?

da-woods commented:
> Wow. I wasn't expecting any further changes in _this_ order in the near future any more (certainly not for 3.1), after introducing the shared module.

I'm completely happy with this missing 3.1 (and think it probably should...).

I suspect it's not as bad as you think performance-wise, but we'll have to see once working properly.

da-woods committed:
Fix utility code passing context to requirements

da-woods commented:
Marking this as draft because I do kind of agree with the scepticism here. At this stage I'd just like to see what the performance impact is (assuming it works), but either way I'd ignore it for 3.1.

da-woods committed:
Merge branch 'master' into shared-abi-refactor-2

da-woods committed:
Maybe fix Limited API

it's impossible to tell; run_benchmarks seems to reliably use
cached C code so it's very hard to test it on modified code.

da-woods committed:
Merge remote-tracking branch 'real_origin/master' into shared-abi-refactor-2
